Ability to abort ongoing tile requests if there are too many

supersonicclay opened the pull request:
<!-- For feature, feature enhancement or bug fix, create an issue first and finish To Do List there -->
<!-- Anything doesn't work as expected is a bug, including code, doc and test -->
Fixes #4831
<!-- For other PRs without open issue -->
#### Background

<!-- For all the PRs -->
#### Change List
- Changed the default `maxRequests` from 8 to 6 to correlate with many browsers' limit for concurrent HTTP requests (https://stackoverflow.com/questions/985431/max-parallel-http-connections-in-a-browser)
- Pass an [AbortSignal](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal) object to `getTileData` calls in case a request is aborted.
- Abort requests for unselected tiles if `# of ongoing requests > maxRequests`

#### Demo

This demonstrates tile requests that each take 5 seconds and specifying `maxRequests=6`. Notice tons of requests are started as the user zooms and pans around. But when they settle the viewport, only the requests for the visible tiles are pending. Before this PR, all the tile requests would be queued up and would have to finish before the requests for the currently visible tiles.

![cancel-tile](https://user-images.githubusercontent.com/469582/90065694-e2bf6800-dca9-11ea-9556-997e6541cb61.gif)

supersonicclay commented on examples/get-started/react/basic/app.js:
```diff
@@ -36,31 +32,99 @@ function Root() {
+          // return new Promise(resolve => setInterval(resolve, wait));
+
+          // docker run --rm -it -p 7000:80 ealen/echo-server
+          await fetch(
```
can ignore changes to this example, it was just an easy way to test my changes for now

kylebarron commented:
Currently the cancellation happens here. When the request scheduler comes to this tile in the queue, it checks if `this.isSelected` is `1`.
https://github.com/visgl/deck.gl/blob/1aff4952f5a5229abcd18e50af977c25eabee76a/modules/geo-layers/src/tile-layer/tile-2d-header.js#L44-L46

`isSelected` is set here:
https://github.com/visgl/deck.gl/blob/1aff4952f5a5229abcd18e50af977c25eabee76a/modules/geo-layers/src/tile-layer/tileset-2d.js#L171-L172

So at this location you could check the existing value of `tile.isSelected`, and if it had previously been selected, but is no longer, you could call `tile.abort()`.

kylebarron commented on modules/geo-layers/src/tile-layer/tileset-2d.js:
```diff
@@ -269,11 +278,22 @@ export default class Tileset2D {
+    for (const tile of this._cache.values()) {
+      // Abort ongoing requests for tiles that are no longer visible
+      // TODO: Make this based on max outstanding requests rather than immediately aborting off-screen requests
+      if (!tile.isVisible && !tile.isLoaded) {
```
I think you want to use `isSelected`, not `isVisible`. The latter is true even if the zoom level has changed dramatically. `isSelected` is true only when the tile is in view and is at the current zoom level.

supersonicclay commented:
@kylebarron I updated how it determines which tiles to abort and when. I introduced a new `maxOngoingRequests` prop. That users can use to control at what point, it'll start aborting tile requests.

It will now look at all the unselected tiles that haven't finished loading. If there are `> maxOngoingRequests`, then it'll abort however many it needs to get back to `<= maxOngoingRequests`. I tried to use the `maxRequests` prop, but it didn't seem to play well with the request scheduler. Also, it seems like `maxRequests` should really be `maxQueuedRequests` since it seems to be limiting the number of queued requests that haven't started yet.

Pessimistress commented:
I think having multiple `max*Requests` props will be confusing to the user. The user's intention in setting `maxRequests` is clear. The implementation conflicts should be resolved internally.

supersonicclay commented:
@kylebarron @Pessimistress this is ready for review again. I cleaned it up and made it work with the existing `maxRequests` parameter and the `RequestScheduler`.

kylebarron commented:
> * Changed the default `maxRequests` from 8 to 6 to correlate with many browsers' limit for concurrent HTTP requests ([stackoverflow.com/questions/985431/max-parallel-http-connections-in-a-browser](https://stackoverflow.com/questions/985431/max-parallel-http-connections-in-a-browser))

The most recent data I can find on that page [suggests](https://stackoverflow.com/a/30064610):

| Browser              | Connections per Domain         | Max Connections                |
| -------------------- | ------------------------------ | ------------------------------ |
| Chrome 81            | 6 [^note1]                     | 256[^note2]                    |
| Edge 18              | *same as Internet Explorer 11* | *same as Internet Explorer 11* |
| Firefox 68           | 9 [^note1] or 6 [^note3]       | 1000+[^note2]                  |
| Internet Explorer 11 | 12 [^note4]                    | 1000+[^note2]                  |
| Safari 13            | 6 [^note1]                     | 1000+[^note2]                  |

This reinforces the value of having a `data` prop as an array of `{a,b,c}.*` domains to get around the per domain limit. So I suppose the OSM example could increase `maxRequests` to `18`, since the OSM tile server has three subdomains. But `6` is a fine default.

kylebarron commented on docs/api-reference/geo-layers/tile-layer.md:
```diff
@@ -114,6 +114,7 @@ If supplied, `getTileData` is called to retrieve the data of each tile. It recei
 - `z` (Number) - z index of the tile
 - `url` (String) - resolved url of the tile if the `data` prop is provided, otherwise `null`
 - `bbox` (Object) - bounding box of the tile. When used with a geospatial view, `bbox` is in the shape of `{west: <longitude>, north: <latitude>, east: <longitude>, south: <latitude>}`. When used with a non-geospatial view, `bbox` is in the shape of `{left, top, right, bottom}`.
+- `signal` (Object) - an [AbortSignal](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal) that may be signalled if there are too many ongoing requests.
```
The `AbortSignal` is only raised if the tile is no longer on the screen...? There will never be too many ongoing `fetch` requests because the `RequestScheduler` prevents `fetch` being called more than 6 times at once...

supersonicclay replied:
I reworded this to "queued" rather than "ongoing".

kylebarron commented on docs/api-reference/geo-layers/tile-layer.md:
```diff
@@ -179,13 +180,15 @@ How the tile layer refines the visibility of tiles. One of the following:
+If `> 0`, a maximum of `maxRequests` instances of `getTileData` will be called concurrently. Requests may never be called if the tile wasn't visible long enough to be scheduled and started. Requests may also be aborted (through the `signal` passed to `getTileData`) if there are more than `maxRequests` ongoing requests and some of those are for tiles that are no longer visible.
 
-- Default: `8`
+If `getTileData` makes `fetch` requests, then `maxRequests` should correlate to the browser's maximum number of concurrent `fetch` requests. For Chrome, the max is 6.
```
I'd specify "6 per domain", and "to increase this limit, utilize the `data` prop to support multiple domains".

kylebarron commented on modules/geo-layers/package.json:
```diff
@@ -39,7 +39,8 @@
     "long": "^3.2.0",
-    "math.gl": "^3.2.1"
+    "math.gl": "^3.2.1",
+    "abortcontroller-polyfill": "^1.5.0"
```
`AbortController` is [supported on everything except IE11](https://caniuse.com/#feat=abortcontroller). Does deck.gl support IE11?

supersonicclay replied:
The tests failed without the polyfill. I'm open to suggestions if there is a structure for polyfilling whatever engine the tests use and I can get rid of this dependency.

kylebarron replied:
Hmm I'm not sure then. Xiaoji probably has a better idea.

Pessimistress replied:
This polyfill should be added to the root package.json as a devDependency, since it's only needed in Node.

supersonicclay replied:
How do you recommend I conditionally include it from node? Something like:
```
if (__NODE__) {
    require('abortcontroller-polyfill');
}
```

supersonicclay replied:
I made an attempt, but it didn't work:
https://travis-ci.com/github/visgl/deck.gl/jobs/372367371
https://github.com/visgl/deck.gl/pull/4838/commits/1805d4e40ae6fa945fde83ed5e6aa9068fd37d7d

Pessimistress replied:
Import it in test/node.js

The published module is not supposed to support node.

Pessimistress replied:
You need to require the polyfill after `fetch` is polyfilled:

https://travis-ci.com/github/visgl/deck.gl/jobs/372367371#L6937

kylebarron commented on modules/geo-layers/src/tile-layer/tileset-2d.js:
```diff
@@ -1,3 +1,5 @@
+/* eslint-env browser */
```
It doesn't look like you use browser APIs in this file?

kylebarron commented on docs/api-reference/geo-layers/tile-layer.md:
```diff
@@ -114,6 +114,7 @@ If supplied, `getTileData` is called to retrieve the data of each tile. It recei
 - `z` (Number) - z index of the tile
 - `url` (String) - resolved url of the tile if the `data` prop is provided, otherwise `null`
 - `bbox` (Object) - bounding box of the tile. When used with a geospatial view, `bbox` is in the shape of `{west: <longitude>, north: <latitude>, east: <longitude>, south: <latitude>}`. When used with a non-geospatial view, `bbox` is in the shape of `{left, top, right, bottom}`.
+- `signal` (Object) - an [AbortSignal](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal) that may be signalled if there are too many queued requests. Note: only tiles that aren't visible will be aborted.
```
It might be good to add in the docs somewhere that a custom `getTileData` can check `signal.aborted` manually. So it can do:

```js
const {signal} = tile;
const data = fetch(url, {signal});

if (signal.aborted) {
   return null;
}
// Expensive computation on returned data
```

I believe with this PR, if a tile is aborted, the response is never cached? This might hurt performance in a situation where the user doesn't implement `signal`, i.e. doesn't pass it to `fetch`. In this case, the entire `fetch` and data processing would still happen but since the `TileLayer` instigated an `abort`, the `TileLayer` wouldn't cache the response. Maybe there's a way to abort the `getTileData`, but _if_ it still returns a non-null response, then cache it?

supersonicclay replied:
Awesome suggestions. Nice.

kylebarron commented on modules/geo-layers/src/tile-layer/tileset-2d.js:
```diff
@@ -170,6 +175,23 @@ export default class Tileset2D {
+
+    if (maxRequests > 0 && ongoingRequestCount > maxRequests) {
+      // There are too many ongoing requests, see if any can be aborted
+      const tilesToAbort = abortCandidates.slice(0, ongoingRequestCount - maxRequests);
```
I don't understand why this would ever be nonzero. I thought the idea of the `RequestScheduler` was to release a max of `maxRequests` at a time from the queue. So if `maxRequests` is 6, there should never be more than 6 tiles at one time where `isLoading` is `true`.

In my mind the preferred implementation is just:
```js
if (tile.isLoading && !tile.isSelected) {
  tile.abort();
}
```

kylebarron replied:
In this case, assuming that the user's `getTileData` implements `signal` checking, then `getTileData` would return immediately and the `finally` block would call `requestToken.done();`, which would then allow the `RequestScheduler` to release a new tile request.

supersonicclay replied:
There could be requests loading but not selected (e.g. user panned a bit and panned back), in this case we wouldn't want to abort them unless we needed to free up some requests.

kylebarron replied:
> user panned a bit and panned back

That's a valid scenario that I also had in mind. I think instead of using `ongoingRequestCount`, you should use [`requestScheduler.requestQueue.length`](https://github.com/visgl/loaders.gl/blob/62e52a6e914f517f1f1da967b9a9bd81f9ea93d5/modules/loader-utils/src/lib/request-utils/request-scheduler.js#L26), which will give the total number of requests that are queued, instead of just the number that have already been started.

supersonicclay replied:
is `requestScheduler.requestQueue.length` the include requests that have started but aren't yet done or just the number that are queued (not started)?

kylebarron replied:
Just abort all tiles out of view if you there's a queue?

```js
const nTilesQueued = this.requestScheduler.requestQueue.length;
if (maxRequests > 0 && nTilesQueued > maxRequests) {
  this._cache
  .values()
  .filter(tile => tile.isLoading && !tile.isSelected)
  .map(tile => tile.abort);
}
```

supersonicclay replied:
We could do that, but we should only abort the number we need to get back under the max. For example, no need to abort all 6 tiles out of view just because one new one came into view and is queue.

supersonicclay replied:
The thing with this approach is that we have to iterate the cache twice. With the code I have in the PR right now, we only iterate the cache once to build up the tiles to abort.

Pessimistress replied:
Without having to create new arrays:

```js
if (maxRequests > 0) {
  while (ongoingRequestCount > maxRequests && abortCandidates.length) {
    const tile = abortCandidates.shift();
    tile.abort();
    ongoingRequestCount--;
  }
}
```

supersonicclay replied:
Made the update. Hopefully this'll work.

supersonicclay commented:
> > * Changed the default `maxRequests` from 8 to 6 to correlate with many browsers' limit for concurrent HTTP requests ([stackoverflow.com/questions/985431/max-parallel-http-connections-in-a-browser](https://stackoverflow.com/questions/985431/max-parallel-http-connections-in-a-browser))
> 
> The most recent data I can find on that page [suggests](https://stackoverflow.com/a/30064610):
> 
> Browser	Connections per Domain	Max Connections
> Chrome 81	6 [^note1]	256[^note2]
> Edge 18	_same as Internet Explorer 11_	_same as Internet Explorer 11_
> Firefox 68	9 [^note1] or 6 [^note3]	1000+[^note2]
> Internet Explorer 11	12 [^note4]	1000+[^note2]
> Safari 13	6 [^note1]	1000+[^note2]
> This reinforces the value of having a `data` prop as an array of `{a,b,c}.*` domains to get around the per domain limit. So I suppose the OSM example could increase `maxRequests` to `18`, since the OSM tile server has three subdomains. But `6` is a fine default.

Found out that when the web server supports HTTP 2, then concurrent request are unbounded since the same TCP connection is used for all the requests. I added more to the docs about it and updated the example to not limit max requests (since the OSM tiles support HTTP 2). The example app is much much faster now too!

kylebarron commented:
Doesn't the client have to support HTTP 2 as well? Is there a worry that a client might not support HTTP 2 and then you'd have a huge queue of requests? I believe Xiaoji had previously warned about having a high number of requests to OSM, but I'll let her chime in.

supersonicclay commented:
I'm fine with whatever max makes sense for the example app. But we are now not restricted to 18.

supersonicclay commented:
Updated the example app to make 20 concurrent requests. It still seems very fast. Let me know what else you want to see. But from my perspective, this should be good to merge.

supersonicclay commented:
Ok, I got the AbortController polyfill working. I think I've addressed all the comments, but let me know if you need anything else with this PR. Thanks for all the review and suggestions!

kylebarron commented on docs/api-reference/geo-layers/tile-layer.md:
```diff
@@ -114,11 +114,24 @@ If supplied, `getTileData` is called to retrieve the data of each tile. It recei
+const data = fetch(url, {signal});
+
+if (signal.aborted) {
+   return null;
```
Maybe add a sentence here saying "If you check for `signal.aborted` and interrupt your processing, you should return `null` so that incorrect data is not cached" ?

supersonicclay replied:
They can return falsy or throw. Either way it will not be cached.

supersonicclay replied:
But I can spell that out in the docs if you think it needs it.

kylebarron replied:
Yes you're right. Maybe "If you abort `getTileData`, don't return incomplete data or it will be cached". I think one such sentence in the docs would be helpful.

kylebarron commented on docs/api-reference/geo-layers/tile-layer.md:
```diff
@@ -179,13 +192,17 @@ How the tile layer refines the visibility of tiles. One of the following:
+If `getTileData` makes `fetch` requests against an HTTP 1 web server, then `maxRequests` should correlate to the browser's maximum number of concurrent `fetch` requests. For Chrome, the max is 6 per domain. If you use the `data` prop and specify multiple domains, you can increase this limit. For example, with Chrome and 3 domains specified, you can set `maxRequests=18`.
 
-If `> 0`, a maximum of `maxRequests` instances of `getTileData` will be called concurrently. Additionally, for requests that are throttled, if the tile is no longer visible in the viewport when a request slot opens up, that request will be cancelled. When requests are not throttled, `getTileData` will be called for all visible tiles, regardless of how long that tile is or was visible.
+If the web server supports HTTP/2 (Open Chrome dev tools and look for "h2" in the Protocol column), then you can make an unlimited number of concurrent requests (and can set `maxRequests=-1`).
```
I might add "Note that this will request data for every tile, no matter for how long the tile was visible, and may increase server load."

kylebarron commented:
Also once this is merged it would be good to add `AbortSignal` support to the MVTLayer and TerrainLayer. Since they override `getTileData` you'd need to change those manually.

Pessimistress commented on modules/core/src/lib/layer.js:
```diff
@@ -65,9 +65,9 @@ const defaultProps = {
+    value: (url, {propName, layer, signal}) => {
       const {resourceManager} = layer.context;
-      const loadOptions = layer.getLoadOptions();
+      const loadOptions = signal ? {...layer.getLoadOptions(), signal} : layer.getLoadOptions();
```
The `signal` parameter is only used by the `TileLayer`, which should not use `resorceManager`.

If we start using AbortSignal as a generic solution in the future, this line still breaks the `resourceManager` logic below.

I recommend overriding the `fetch` implementation in `TileLayer.defaultProps` to avoid unexpected breakage.

supersonicclay replied:
Can we do this piece in a separate PR? I think this one is separate.

supersonicclay replied:
I can make the change. Which should it be?
```
TileLayer.defaultProps.fetch = window.fetch;
```
or
```
import {load} from '@loaders.gl/core';
TileLayer.defaultProps.fetch = load;
```

supersonicclay replied:
I assume the latter and updated this PR.

kylebarron replied:
Check out the definition in the base layer:
https://github.com/visgl/deck.gl/blob/c8702ae134e0598b2fdca03642744f25e9de593b/modules/core/src/lib/layer.js#L66-L90

supersonicclay replied:
Made the update to look more like core layer's fetch, but without the resource manager parts.

supersonicclay commented:
Let me know what else you need before this can be merged. I believe I took care of the asks. Thanks.

ilan-gold commented:
Is an alpha release possible for this?  Very excited to integrate this / be a guinea pig for testing this!

kylebarron commented:
Yeah I'd love an alpha release to test this as well. Not sure other maintainers' plans on alpha releases.

ilan-gold commented:
Just an update here - I've tried this and it looks good.  @kylebarron I've got a PR in the works for geotiff to support this.
